spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/BlockingQueueConsumer.java

```diff
@@ -952,6 +950,52 @@ public void handleDelivery(String consumerTag, Envelope envelope, AMQP.BasicProp
 
 	}
 
+	private static final class ConsumerDecorator implements Consumer {
+
+		private final String queue;
+
+		private final Consumer delegate;
+
+		private final ApplicationEventPublisher applicationEventPublisher;
+
+		ConsumerDecorator(String queue, Consumer delegate, ApplicationEventPublisher applicationEventPublisher) {
+			this.queue = queue;
+			this.delegate = delegate;
+			this.applicationEventPublisher = applicationEventPublisher;
+		}
+
+
+		public void handleConsumeOk(String consumerTag) {
+			this.delegate.handleConsumeOk(consumerTag);
+			if (this.applicationEventPublisher != null) {
+				this.applicationEventPublisher.publishEvent(new ConsumeOkEvent(this.delegate, this.queue, consumerTag));
+			}
+		}
+
+		public void handleShutdownSignal(String consumerTag, ShutdownSignalException sig) {
+			this.delegate.handleShutdownSignal(consumerTag, sig);
+		}
+
+		public void handleCancel(String consumerTag) throws IOException {
+			this.delegate.handleCancel(consumerTag);
+		}
+
+		public void handleCancelOk(String consumerTag) {
+			this.delegate.handleCancelOk(consumerTag);
+		}
```

Comment:
Perhaps we should also store the tag in a field; if only so that it's visible in a debugger (save having to look it up in the map). Given that the tag in the delegate gets overwritten when there are multiple queues.

Also implement `toString()` with the queue and tag (again, just for a debugger).